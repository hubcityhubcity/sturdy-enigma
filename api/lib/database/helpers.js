const fetch = require('node-fetch')
const escape = require('escape-html')

const graphQLQuery = ({ hasura_url, hasura_admin_key }) => async (query) => {
  const data = await fetch(hasura_url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-hasura-admin-secret': hasura_admin_key
    },
    body: JSON.stringify({
      query
    })
  }).then((res) => res.json())

  if (data.errors || data.error) {
    console.error('Query failed:', query)
    console.error(
      'GraphQL query failed:',
      JSON.stringify(data.error || data.errors, undefined, 2)
    )
    throw new Error('GraphQL query failed')
  }

  return data
}

function gqlKeyValue([key, value]) {
  // escape all values as they could be malicious
  return `${key}: "${escape(value)}"`
}

// stringify a set of row to be according to the graphQL schema
function stringifyForGraphQL(rows, height, chainId) {
  if (!Array.isArray(rows)) {
    rows = [rows]
  }
  // set the height for every row if specified
  if (height !== undefined) {
    rows = rows.map((row) => Object.assign(row, { height }))
  }
  // set the height for every row if specified
  if (chainId !== undefined) {
    rows = rows.map((row) => Object.assign(row, { chain_id: chainId }))
  }
  return `[
    ${rows
      .map((row) => {
        return `{${Object.entries(row).map(gqlKeyValue).join(', ')}}`
      })
      .join(',\n')}
    ]`
}

// get all columns for a db object
function getColumns(objects) {
  const obj = Array.isArray(objects) ? objects[0] : objects
  return Object.keys(obj)
}

const insert = ({ hasura_url, hasura_admin_key }, upsert = false) => (
  schema
) => async (table, rows, height, chainId, returnKeys) => {
  if ((Array.isArray(rows) && rows.length === 0) || !rows) {
    return
  }

  returnKeys = Array.isArray(returnKeys) ? returnKeys : []
  let schema_prefix = schema ? schema + '_' : ''

  const query = `
        mutation {
            insert_${schema_prefix}${table} (
                objects: ${stringifyForGraphQL(rows, height, chainId)}${
    upsert
      ? `,
                            on_conflict: {
                                constraint: ${table}_pkey,
                                update_columns: [${getColumns(rows)}]
                            }
                            `
      : ''
  }
            )
            {
                affected_rows
                ${
                  returnKeys.length !== 0
                    ? `returning {
                      ${returnKeys.join('\n')}
                    }`
                    : ''
                }
            }
    }
    `
  return graphQLQuery({ hasura_url, hasura_admin_key })(query)
}

const read = ({ hasura_url, hasura_admin_key }) => (schema) => async (
  table,
  queryName,
  keys,
  filter
) => {
  keys = Array.isArray(keys) ? keys : [keys]
  // schema could be set or not
  let schema_prefix = schema ? schema + '_' : ''

  const query = `
        query ${schema_prefix}${queryName} {
            ${schema_prefix}${table}${filter ? `(${filter})` : ''} {
                ${keys.join('\n')}
            }
        }
    `

  const res = await graphQLQuery({ hasura_url, hasura_admin_key })(query)
  return res.data[`${schema_prefix}${table}`]
}

module.exports = {
  insert,
  read,
  query: graphQLQuery
}
