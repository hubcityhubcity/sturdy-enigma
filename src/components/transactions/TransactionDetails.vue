<template>
  <div class="tx__content">
    <component
      :is="messageTypeComponent"
      :transaction="transaction"
      :validators="validators"
      :session-address="address"
      class="tx__content__left"
    />
    <TransactionMetadata
      v-if="showMetaData"
      :fees="transaction.fees"
      :block="transaction.blockNumber"
      :time="transaction.time"
      class="tx__content__right"
    />
  </div>
</template>

<script>
import { messageType } from "./messageTypes.js"
import TransactionMetadata from "./TransactionMetadata"

import {
  SendMessageDetails,
  DelegateMessageDetails,
  DepositMessageDetails,
  EditValidatorMessageDetails,
  SubmitProposalMessageDetails,
  UndelegateMessageDetails,
  UnjailMessageDetails,
  CreateValidatorMessageDetails,
  VoteMessageDetails,
  BeginRedelegateMessageDetails,
  SetWithdrawAddressMessageDetails,
  WithdrawDelegationRewardMessageDetails,
  WithdrawValidatorCommissionMessageDetails
} from "./message-view"

import Bech32 from "common/Bech32"

export default {
  name: `transaction-details`,
  components: {
    TransactionMetadata,
    Bech32,
    SendMessageDetails,
    DelegateMessageDetails,
    DepositMessageDetails,
    EditValidatorMessageDetails,
    SubmitProposalMessageDetails,
    UndelegateMessageDetails,
    UnjailMessageDetails,
    CreateValidatorMessageDetails,
    VoteMessageDetails,
    BeginRedelegateMessageDetails,
    SetWithdrawAddressMessageDetails,
    WithdrawDelegationRewardMessageDetails,
    WithdrawValidatorCommissionMessageDetails
  },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    validators: {
      type: Object,
      required: true
    },
    address: {
      type: String,
      required: false,
      default: ""
    },
    showMetaData: {
      type: Boolean,
      default: true
    }
  },
  computed: {
    messageTypeComponent: function() {
      // TODO Could improve this using dynamic loading.
      switch (this.transaction.type) {
        case messageType.SEND:
          return `send-message-details`
        case messageType.DELEGATE:
          return `delegate-message-details`
        case messageType.CREATE_VALIDATOR:
          return `create-validator-message-details`
        case messageType.EDIT_VALIDATOR:
          return `edit-validator-message-details`
        case messageType.UNDELEGATE:
          return `undelegate-message-details`
        case messageType.BEGIN_REDELEGATE:
          return `begin-redelegate-message-details`
        case messageType.UNJAIL:
          return `unjail-message-details`
        case messageType.SUBMIT_PROPOSAL:
          return `submit-proposal-message-details`
        case messageType.DEPOSIT:
          return `deposit-message-details`
        case messageType.VOTE:
          return `vote-message-details`
        case messageType.SET_WITHDRAW_ADDRESS:
          return `set-withdraw-address-message-details`
        case messageType.WITHDRAW_DELEGATION_REWARD:
          return `withdraw-delegation-reward-message-details`
        case messageType.WITHDRAW_VALIDATOR_COMMISSION:
          return `withdraw-validator-commission-message-details`
        default:
          return ``
      }
    }
  }
}
</script>

<style scoped>
.tx__content {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  padding: 1rem;
}

.tx__content__left,
.tx__content__right {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.tx__content__right {
  text-align: right;
}

.tx__content__information,
.tx__content__information > * {
  display: flex;
  flex-direction: row;
}

.tx__content__information,
.tx__content__right {
  font-size: 14px;
  color: var(--dim);
}

.tx__content__caption {
  line-height: 18px;
  font-size: 18px;
  color: var(--bright);
}

@media screen and (max-width: 767px) {
  .tx__content {
    flex-direction: column;
    text-align: left;
  }

  .tx__content__right {
    text-align: left;
    padding-top: 0.5rem;
  }
}
</style>
