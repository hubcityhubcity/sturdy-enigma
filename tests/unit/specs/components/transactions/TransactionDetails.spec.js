import { mount, createLocalVue } from "@vue/test-utils"
import TransactionDetails from "src/components/transactions/TransactionDetails"
import { testTransactionObjects } from "tests/unit/helpers/testValues"
describe(`TransactionDetails`, () => {
  let wrapper

  const localVue = createLocalVue()
  localVue.directive(`tooltip`, () => {})
  localVue.directive(`clipboard`, () => {})

  beforeEach(() => {
    wrapper = mount(TransactionDetails, {
      localVue,
      propsData: {
        transaction: testTransactionObjects[0],
        validators: {},
        address: "cosmos1"
      },
      stubs: ["router-link"]
    })
  })
  it(`renders MsgSend transaction item unrelated to session user`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgSend transaction item from session user to other`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[1] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgSend transaction item to yourself`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[2] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgSend transaction item from other to you`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[3] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgUndelegate transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[4] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgBeginRedelegate transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[5] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgDelegate transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[6] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgWithdrawDelegationReward transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[7] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgSubmitProposal transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[8] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgVote transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[9] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgDeposit transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[10] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgSetWithdrawAddress transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[11] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgSetWithdrawValidatorCommission transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[12] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgUnjail transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[13] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgCreateValidator transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[14] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`renders MsgEditValidator transaction item`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[15] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`Unknown transaction returns empty string`, () => {
    wrapper.setProps({
      transaction: { ...testTransactionObjects[15], type: "Unknown" }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
