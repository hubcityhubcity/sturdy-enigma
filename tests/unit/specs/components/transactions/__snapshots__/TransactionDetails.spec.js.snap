// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionDetails Unknown transaction returns empty string 1`] = `
<div
  class="tx__content"
>
  <!---->
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgBeginRedelegate transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Redelegated
      
        <b>
          0.01
        </b>
         
        <span>
           ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    From 
    
      <router-link-stub
        to="staking/validators/cosmosvaloper1sjllsnasdmasakjshdlkdhjasldnasd"
      >
        
      cosmosvaloper…nasd 
      </router-link-stub>
      To 
    
      <router-link-stub
        to="staking/validators/cosmosvaloper1uaisjdnidjnlaskjdnlkjasndlkadjn"
      >
        
      cosmosvaloper…adjn
    
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        20.00083
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #726616 
      </router-link-stub>
      
    @ Jun 19th 2019 14:45:43
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgCreateValidator transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Create validator
    
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    Monikor 
    
      <router-link-stub
        to="staking/validators/cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkj"
      >
        
      cosmos…dlkj
    
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgDelegate transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Delegated
      
        <b>
          0.05
        </b>
         
        <span>
           ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    To 
    
      <router-link-stub
        to="staking/validators/cosmosvaloper1uaisjdnidjnlaskjdnlkjasndlkadjn"
      >
        
      cosmosvaloper…adjn
    
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        10.00066
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #726599 
      </router-link-stub>
      
    @ Jun 19th 2019 14:43:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgDeposit transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Deposited
      
        <b>
          0.2345
        </b>
         
        <span>
          ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    On
    
      <router-link-stub
        to="/governance/2"
      >
        Proposal #2
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        69
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626103 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgEditValidator transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        Edit validator
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    Monikor 
    
      <router-link-stub
        to="staking/validators/cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkj"
      >
        
      cosmos…dlkj
    
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgSend transaction item from other to you 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    validators="[object Object]"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Received
      
        <b>
           0.2345
        </b>
         
        <span>
           ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
      From 
      
      <div
        class="bech32-address"
      >
        <div
          class="address"
        >
          
    cosmos…1212
  
        </div>
         
        <div
          class="copied"
        >
          <i
            class="material-icons"
          >
            check
          </i>
        </div>
      </div>
       
      <span>
         - (Sent via Lunie)
      </span>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0.00037
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #1086769 
      </router-link-stub>
      
    @ Jul 31st 2019 09:22:23
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgSend transaction item from session user to other 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    validators="[object Object]"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Sent
      
        <b>
           0.2345
        </b>
         
        <span>
           ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
      To 
      
      <div
        class="bech32-address"
      >
        <div
          class="address"
        >
          
    cosmos…sdas
  
        </div>
         
        <div
          class="copied"
        >
          <i
            class="material-icons"
          >
            check
          </i>
        </div>
      </div>
       
      <span>
         - (Sent via Lunie)
      </span>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0.00037
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #1086769 
      </router-link-stub>
      
    @ Jul 31st 2019 09:22:23
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgSend transaction item to yourself 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    validators="[object Object]"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Sent
      
        <b>
           0.2345
        </b>
         
        <span>
           ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
      To yourself!
     
      <span>
         - (Sent via Lunie)
      </span>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0.00037
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #1086769 
      </router-link-stub>
      
    @ Jul 31st 2019 09:22:23
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgSend transaction item unrelated to session user 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    validators="[object Object]"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Sent
      
        <b>
           0.2345
        </b>
         
        <span>
           ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
      From 
      
      <div
        class="bech32-address"
      >
        <div
          class="address"
        >
          
    cosmos…dlkj
  
        </div>
         
        <div
          class="copied"
        >
          <i
            class="material-icons"
          >
            check
          </i>
        </div>
      </div>
       to 
      
      <div
        class="bech32-address"
      >
        <div
          class="address"
        >
          
    cosmos…dlkj
  
        </div>
         
        <div
          class="copied"
        >
          <i
            class="material-icons"
          >
            check
          </i>
        </div>
      </div>
       
      <span>
         - (Sent via Lunie)
      </span>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0.00037
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #1086769 
      </router-link-stub>
      
    @ Jul 31st 2019 09:22:23
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgSetWithdrawAddress transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        Update withdraw address
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    To cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkj
  
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgSetWithdrawValidatorCommission transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        Withdraw validator commission
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    From 
    
      <router-link-stub
        to="staking/validators/cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkj"
      >
        cosmos…dlkj
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgSubmitProposal transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Submitted text proposal
      
        <b>
          0.2
        </b>
         
        <span>
           ATOM
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    Title 
      <i>
        Test title
      </i>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        69
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgUndelegate transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        
      Undelegated
      
        <b>
          0.05
        </b>
         
        <span>
           ATOM
        </span>
         
        <span
          class="tx-unbonding__time-diff"
        >
           (liquid in 50 years)
        </span>
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    From 
    
      <router-link-stub
        to="/validators/cosmosvaloper1lsajkdalskdjhasldkahsdljkalskdjhlaks"
      >
        
      cosmosvaloper…laks
    
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        410.00041
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #1248479 
      </router-link-stub>
      
    @ Jul 31st 2019 09:22:23
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgUnjail transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        Unjail
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    Name:
    
      <router-link-stub
        to="staking/validators/cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkj"
      >
        cosmos…dlkj
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgVote transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        Voted NoWithVeto
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    On 
    
      <router-link-stub
        to="/governance/13"
      >
        Proposal #13
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        69
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626103 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;

exports[`TransactionDetails renders MsgWithdrawDelegationReward transaction item 1`] = `
<div
  class="tx__content"
>
  <div
    class="tx__content__left"
    session-address="cosmos1"
  >
    <div
      class="tx__content__caption"
    >
      <p>
        Withdrawal
      </p>
    </div>
     
    <div
      class="tx__content__information"
    >
      
    From 
    
      <router-link-stub
        to="staking/validators/cosmosvaloper1sjllsnasdmasakjshdlkdhjasldnasdnadn"
      >
        cosmosvaloper…nadn
      </router-link-stub>
    </div>
  </div>
   
  <div
    class="tx__content__right"
  >
    <div>
      
    Network Fee: 
    
      <b>
        0
      </b>
       
      <span>
         ATOM
      </span>
    </div>
     
    <div>
      <router-link-stub
        to="[object Object]"
      >
        Block #626102 
      </router-link-stub>
      
    @ Jun 11th 2019 13:47:45
  
    </div>
  </div>
</div>
`;
